Fix: Remove balance check from EstimateSwap

mattverse opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #2337 

## What is the purpose of the change

EstimateSwap currently checks the balance of the query sender, not because it actually moves the assets on an actual swap, but beacuse it uses the same API with the actual multi swaps that happen. 

This PR separates out the API needed for the estimation so that it does not check balance of the query sender. 

(Special thanks to @vuong177 the original author of this PR!)


## Brief Changelog

- Remove balance check from `EstimateSwap`

## Testing and Verifying
This change added tests. 


## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.mxd`? yes
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

mattverse commented:
Go Test CI is failing rn because we need a new wasm reflect binary

ValarDragon commented on wasmbinding/message_plugin.go:
```diff
@@ -218,50 +218,22 @@ func PerformSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.Ac
-		routes := []gammtypes.SwapAmountInRoute{{
-			PoolId:        swap.First.PoolId,
-			TokenOutDenom: swap.First.DenomOut,
-		}}
-		for _, step := range swap.Route {
-			routes = append(routes, gammtypes.SwapAmountInRoute{
-				PoolId:        step.PoolId,
-				TokenOutDenom: step.DenomOut,
-			})
-		}
-		if swap.Amount.ExactIn.Input.IsNegative() {
-			return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap negative amount in"}
-		}
-		tokenIn := sdk.Coin{
-			Denom:  swap.First.DenomIn,
-			Amount: swap.Amount.ExactIn.Input,
+		routes, tokenIn, tokenOutMinAmount, err := getSwapExactAmountInParams(swap)
+		if err != nil {
+			return nil, err
 		}
-		tokenOutMinAmount := swap.Amount.ExactIn.MinOutput
```
Confirmed this is just a move

ValarDragon commented on wasmbinding/message_plugin.go:
```diff
@@ -218,50 +218,22 @@ func PerformSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.Ac
 		}
 		return &bindings.SwapAmount{Out: &tokenOutAmount}, nil
 	} else if swap.Amount.ExactOut != nil {
-		routes := []gammtypes.SwapAmountOutRoute{{
-			PoolId:       swap.First.PoolId,
-			TokenInDenom: swap.First.DenomIn,
-		}}
-		output := swap.First.DenomOut
-		for _, step := range swap.Route {
-			routes = append(routes, gammtypes.SwapAmountOutRoute{
-				PoolId:       step.PoolId,
-				TokenInDenom: output,
-			})
-			output = step.DenomOut
-		}
-		tokenInMaxAmount := swap.Amount.ExactOut.MaxInput
-		if swap.Amount.ExactOut.Output.IsNegative() {
-			return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap negative amount out"}
-		}
-		tokenOut := sdk.Coin{
-			Denom:  output,
-			Amount: swap.Amount.ExactOut.Output,
```
confirmed just a move

ValarDragon reviewed (commented):
Reviewed all of the wasmbinding and proto changes, looks great to me! Reviewing keeper logic now

ValarDragon commented on x/gamm/keeper/estimate_swap.go:
```diff
@@ -0,0 +1,256 @@
+			"can't get more tokens out than there are tokens in the pool")
+	}
+
+	tokenIn, err = pool.SwapInAmtGivenOut(ctx, sdk.Coins{tokenOut}, tokenInDenom, swapFee)
```
Oh hrm, this isn't so simple.

Ok I think the current method is actually better. Where the thinking is "this state may be  altered, but its up to the caller to ensure this updated state does not get saved (e.g. is in a cache ctx)"

The alternative to get code-reuse would be a deeper refactor, which does seem a bit inadvisable for this PR / release

mattverse replied:
yeah exactly!

ValarDragon commented:
(Had a call to rename the internal `estimate` methods to better indicate that they can really change state, and move `cacheCtx` from query into the go API for public Estimate methods

mattverse commented on x/gamm/keeper/estimate_swap.go:
```diff
@@ -188,68 +154,11 @@ func (k Keeper) EstimateSwapExactAmountOut(
-// estimateUpdatePoolForSwap updates the pool balance.
-// This method does not include emitting hooks or sending the actual asset from
-// the sender to the pool or vice versa, as this is designed to be used by query methods.
-func (k Keeper) estimateUpdatePoolForSwap(
```
changed to method `updatePoolForSwap` in swap.go, as this should also not be an estimating method

p0mvn commented on wasmbinding/queries.go:
```diff
@@ -95,19 +95,48 @@ func (qp QueryPlugin) EstimateSwap(ctx sdk.Context, estimateSwap *bindings.Estim
 	return estimate, err
 }
 
+// EstimatePerformSwap can be used to query
```
nit: unfinished sentence

p0mvn commented on x/gamm/keeper/grpc_query_test.go:
```diff
@@ -283,3 +283,124 @@ func (suite *KeeperTestSuite) TestQueryBalancerPoolSpotPrice() {
+		result    sdk.Int
+	}{
+		{
+			name: "multi route",
```
non-blocking comment / feel free to make an issue if you agree: only testing happy path here, we might want to test error cases as well

mattverse replied:
added test cases for empty requests, I don't think error catching paths for estimate swap itself have to go here

p0mvn commented on x/gamm/keeper/estimate_swap.go:
```diff
@@ -0,0 +1,164 @@
+
+// EstimateMultihopSwapExactAmountIn iterates `EstimateSwapExactAmountIn` and returns
+// the total token out given route for multihop.
+func (k Keeper) EstimateMultihopSwapExactAmountIn(
```
There seems to be a lot of code copied over from non-estimate versions of this method. I'm wondering if it's possible to generalize this somehow so that we are not repeating so much boilerplate.

mattverse replied:
It ranges over the same for loop for the multihops, the only difference is that it would call different internal methods, `swapExactAmountOut` for the actual swap and`estimateSwapExactAmountOut` for the estimated multihops.

This could be solved by passing in the internal func we want to use as a function parameter, but I think such refactor could be done in a separate PR

mattverse replied:
https://github.com/osmosis-labs/osmosis/issues/3386

p0mvn commented on x/gamm/keeper/swap.go:
```diff
@@ -45,35 +45,59 @@ func (k Keeper) swapExactAmountIn(
+) (updatedPool types.PoolI, tokenOut sdk.Coin, err error) {
 	if tokenIn.Denom == tokenOutDenom {
-		return sdk.Int{}, errors.New("cannot trade same denomination in and out")
+		return pool, sdk.Coin{}, errors.New("cannot trade same denomination in and out")
```
```suggestion
		return nil, sdk.Coin{}, errors.New("cannot trade same denomination in and out")
```

p0mvn commented on x/gamm/keeper/swap.go:
```diff
@@ -45,35 +45,59 @@ func (k Keeper) swapExactAmountIn(
 	tokenOutCoin, err := pool.SwapOutAmtGivenIn(ctx, tokensIn, tokenOutDenom, swapFee)
 	if err != nil {
-		return sdk.Int{}, err
+		return pool, sdk.Coin{}, err
```
```suggestion
		return nil, sdk.Coin{}, err
```

p0mvn commented on x/gamm/keeper/swap.go:
```diff
@@ -45,35 +45,59 @@ func (k Keeper) swapExactAmountIn(
 
 	if !tokenOutAmount.IsPositive() {
-		return sdk.Int{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount must be positive")
+		return pool, sdk.Coin{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount must be positive")
```
```suggestion
		return nil, sdk.Coin{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount must be positive")
```

p0mvn commented on x/gamm/keeper/swap.go:
```diff
@@ -45,35 +45,59 @@ func (k Keeper) swapExactAmountIn(
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+) (updatedPool types.PoolI, tokenOut sdk.Coin, err error) {
```
Do we even need to return `updatedPool` here? `pool` is a pointer so it should be mutated if that's desired

mattverse replied:
ur right, they do get mutated if desired, removing returning the updatedPool

p0mvn commented on x/gamm/keeper/swap.go:
```diff
@@ -105,44 +129,66 @@ func (k Keeper) swapExactAmountOut(
+	tokenInMaxAmount sdk.Int,
+	tokenOut sdk.Coin,
+	swapFee sdk.Dec,
+) (updatedPool types.PoolI, tokenIn sdk.Coin, err error) {
```
Same question about `updatedPool`

p0mvn commented on x/gamm/keeper/swap.go:
```diff
@@ -105,44 +129,66 @@ func (k Keeper) swapExactAmountOut(
+) (updatedPool types.PoolI, tokenIn sdk.Coin, err error) {
 	if tokenInDenom == tokenOut.Denom {
-		return sdk.Int{}, errors.New("cannot trade same denomination in and out")
+		return pool, sdk.Coin{}, errors.New("cannot trade same denomination in and out")
```
We should also consider returning `nil` in all error cases

czarcas7ic commented on x/gamm/keeper/estimate_swap.go:
```diff
@@ -0,0 +1,200 @@
+// This method does not execute the full steps of an actaul swap, nor does it change the state after estimation.
+// Returns the estimated amount of token out using cache context.
+func (k Keeper) EstimateSwapExactAmountIn(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+) (sdk.Int, error) {
+	return k.estimateSwapExactAmountIn(ctx, poolId, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee, true)
+}
+
+// estimateSwapExactAmountIn returns the amount of token out given the amount of token in.
+// This method should only be called by estimate methods, as this method does not include full steps of an actual swap.
+func (k Keeper) estimateSwapExactAmountIn(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+	useCacheCtx bool,
+) (sdk.Int, error) {
+	if useCacheCtx {
+		ctx, _ = ctx.CacheContext()
+	}
+	pool, err := k.getPoolForSwap(ctx, poolId)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	tokenOut, err := k.swapExactAmountInNoTokenSend(ctx, pool, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee)
+	if err != nil {
+		return tokenOut.Amount, err
+	}
+
+	err = k.updatePoolForSwap(ctx, pool, tokenIn, tokenOut)
+	return tokenOut.Amount, err
+}
```
Can we make it clearer what the diff is between these two? Both say "this method does not include full steps of an actual swap", but it seems like if cache context is false then we actually mutate state

mattverse replied:
Ahh, yeah this comment does need some better explanation. Added some more in line comments on this, tell me if it makes more sense now!

czarcas7ic commented on x/gamm/keeper/estimate_swap.go:
```diff
@@ -0,0 +1,200 @@
+	cacheCtx, _ := ctx.CacheContext()
+	swapFeeMultiplier := sdk.OneDec()
+
+	if types.SwapAmountOutRoutes(routes).IsOsmoRoutedMultihop() {
+		swapFeeMultiplier = types.MultihopSwapFeeMultiplierForOsmoPools.Clone()
+	}
+
+	// Determine what the estimated input would be for each pool along the multihop route
+	insExpected, err := k.createMultihopExpectedSwapOuts(cacheCtx, routes, tokenOut, swapFeeMultiplier)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+	if len(insExpected) == 0 {
+		return sdk.Int{}, nil
+	}
+
+	insExpected[0] = tokenInMaxAmount
+
```
Same about these needing to change with this PR https://github.com/osmosis-labs/osmosis/pull/3352

mattverse replied:
Merged this with the most recent main and then made the according changes, can you review it and confirm?

czarcas7ic commented on x/gamm/keeper/estimate_swap.go:
```diff
@@ -0,0 +1,200 @@
+	return tokenInAmount, nil
+}
+
+// EstimateSwapExactAmountOut estimates the amount of token in given the exact amount of token out.
+// This method does not execute the full steps of an actaul swap, nor does it change the state after estimation.
+// Returns the estimated amount of token out using cache context.
+func (k Keeper) EstimateSwapExactAmountOut(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenInDenom string,
+	tokenInMaxAmount sdk.Int,
+	tokenOut sdk.Coin,
+	swapFee sdk.Dec,
+) (tokenInAmount sdk.Int, err error) {
+	return k.estimateSwapExactAmountOut(ctx, poolId, tokenInDenom, tokenInMaxAmount, tokenOut, swapFee, true)
+}
+
+// estimateSwapExactAmountOut returns the amount of token in given the amount of token out.
+// This method should only be called by estimate methods, as this method does not include full steps of an actual swap.
+func (k Keeper) estimateSwapExactAmountOut(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenInDenom string,
+	tokenInMaxAmount sdk.Int,
+	tokenOut sdk.Coin,
+	swapFee sdk.Dec,
+	useCacheCtx bool,
+) (tokenInAmount sdk.Int, err error) {
+	if useCacheCtx {
+		ctx, _ = ctx.CacheContext()
+	}
+	pool, err := k.getPoolForSwap(ctx, poolId)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
```
Same point about comments for cacheCtx

mattverse replied:
ditto for the comment above

czarcas7ic commented on x/gamm/keeper/estimate_swap_test.go:
```diff
@@ -0,0 +1,345 @@
+				test.param.tokenOutMinAmount)
+			suite.Require().NoError(err)
+
+			// assure that the token out amount is same
```
```suggestion
			// ensure that the token out amount is same
```

czarcas7ic reviewed (commented):
Left some comments, main thing is changing the multi-hop logic to match the new design here https://github.com/osmosis-labs/osmosis/pull/3352

Also, unclear what is going on with the proto gen CI check

mattverse commented on wasmbinding/message_plugin.go:
```diff
@@ -218,50 +218,22 @@ func PerformSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.Ac
+
 		tokenOutAmount, err := keeper.MultihopSwapExactAmountIn(ctx, contractAddr, routes, tokenIn, tokenOutMinAmount)
 		if err != nil {
 			return nil, sdkerrors.Wrap(err, "gamm perform swap exact amount in")
```
lol I actually do agree with you that its pretty unclear, but also didnt want to touch anything outside the scope of this PR. But sinced its addressed, changing this to something more intuitive

mattverse commented on wasmbinding/queries.go:
```diff
@@ -95,19 +95,49 @@ func (qp QueryPlugin) EstimateSwap(ctx sdk.Context, estimateSwap *bindings.Estim
+	if swap == nil {
+		return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap null swap"}
+	}
+	if swap.Amount.ExactIn != nil {
```
Why do you think that's the case? AFAIU, i think `swap.Amount.ExactIn != nil && swap.Amount.ExactOut == nil` can be a possible scenario, for example, when a pool is 1:1 ratio(10 uosmo, 10 uatom), you could be expecting 1 in 1out or similar

czarcas7ic replied:
Right, not only CAN it be a possible scenario, it should be the only scenario for the one sided argument. What I am trying to prevent here is if someone provides a SwapMsg with both a swap.Amount.ExactIn and swap.Amount.ExactOut, instead of erroring they would go through the first for loop. EstimatePerformSwap should only return a valid response if one side is given, not both sides

czarcas7ic replied:
AFAIU, we should never call an estimate method with both the in and the out. We either give it the in to calculate the out or we give it the out to calculate the in

czarcas7ic replied:
I still feel this is the case

mattverse commented on x/gamm/keeper/estimate_swap.go:
```diff
@@ -0,0 +1,200 @@
+		if types.SwapAmountInRoutes(routes).IsOsmoRoutedMultihop() {
+			swapFeeMultiplier = types.MultihopSwapFeeMultiplierForOsmoPools.Clone()
+		}
+
```
Merged this with the most recent main and then made the according changes, can you review it and confirm?

czarcas7ic commented:
Wait, but wouldn't it be as simple as adding a "useCacheCtx" boolean here https://github.com/osmosis-labs/osmosis/blob/bef8a4b3b107b383676d953a9e89e409b2e4f828/x/gamm/keeper/multihop.go#L13-L19 and then we could get rid of the entire estimate_swap.go file?

mattverse commented:
@czarcas7ic I actaully think that's a great idea. Removed the whole file of estimate swap in my last commit and changed the existing swap methods to have a boolean filed for the estimate swaps.

I think this requires detailed, in-depth review as it touches the core swap logic

czarcas7ic commented:
@mattverse will review in depth tomorrow!

p0mvn commented:
Is this v13 blocking?

p0mvn commented:
NVM, I see that we went with the code duplication approach for safety here: https://github.com/osmosis-labs/osmosis/pull/3424

ValarDragon commented:
@mattverse should we close this PR, and re-PR with elements you want added in still?

ValarDragon commented:
Closing per last comment
